File: solution/1823. Find the Winner of the Circular Game/1823_01.cpp
class Solution {
public:
    int findTheWinner(int n, int k) {
        if (k == 1) return n;
        
        vector<int> next(n);
        for (int i = 0; i < n - 1; ++i)
            next[i] = i + 1;
        next[n - 1] = 0;
        
        int cnt = n, i = 0;
        while (cnt--) {
            int pre = i;
            for (int j = 0; j < k - 1; ++j) {
                pre = i;
                i = next[i];
            }
            next[pre] = next[i];
            i = next[i];
        }
        return i + 1;
    }
};


// 0 1 2
// 1 2 0

File: solution/1823. Find the Winner of the Circular Game/1823_01.py
class Solution:
    def findTheWinner(self, n: int, k: int) -> int:
        if k == 1: return n
        
        nxt = [0] * n
        for i in range(n - 1):
            nxt[i] = i + 1
        nxt[n - 1] = 0
        
        i = 0
        while n > 0:
            for _ in range(k - 1):
                pre = i
                i = nxt[i]
            nxt[pre] = nxt[i]
            i = nxt[i]
            n -= 1
        
        return i + 1
